File: sdk/typescript/src/api/devices.ts
import type { HttpTransport } from '../transport.js';
import type { Device } from '../types.js';

export class DevicesApi {
  constructor(private readonly http: HttpTransport) {}

  async list(): Promise<Device[]> {
    const data = await this.http.execute<{ devices: Device[] }>(/* GraphQL */ `
      query Devices {
        devices {
          id name host ip port service app isConnected
          baseUrl isCastDevice isSourceDevice isCurrentDevice
        }
      }
    `);
    return data.devices;
  }

  async get(id: string): Promise<Device | null> {
    const data = await this.http.execute<{ device: Device | null }>(/* GraphQL */ `
      query Device($id: String!) {
        device(id: $id) {
          id name host ip port service app isConnected
          baseUrl isCastDevice isSourceDevice isCurrentDevice
        }
      }
    `, { id });
    return data.device;
  }

  async connect(id: string): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation ConnectDevice($id: String!) { connect(id: $id) }
    `, { id });
  }

  async disconnect(id: string): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation DisconnectDevice($id: String!) { disconnect(id: $id) }
    `, { id });
  }
}

File: sdk/python/src/rockbox_sdk/api/devices.py
"""Cast / source device discovery and control."""

from __future__ import annotations

from ..transport import HttpTransport
from ..types import Device

_FIELDS = (
    "id name host ip port service app isConnected "
    "baseUrl isCastDevice isSourceDevice isCurrentDevice"
)


class DevicesApi:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    async def list(self) -> list[Device]:
        data = await self._http.execute(
            f"query Devices {{ devices {{ {_FIELDS} }} }}"
        )
        return [Device.model_validate(d) for d in data.get("devices", [])]

    async def get(self, id: str) -> Device | None:
        data = await self._http.execute(
            f"query Device($id: String!) {{ device(id: $id) {{ {_FIELDS} }} }}",
            {"id": id},
        )
        raw = data.get("device")
        return Device.model_validate(raw) if raw is not None else None

    async def connect(self, id: str) -> None:
        await self._http.execute(
            "mutation ConnectDevice($id: String!) { connect(id: $id) }", {"id": id}
        )

    async def disconnect(self, id: str) -> None:
        await self._http.execute(
            "mutation DisconnectDevice($id: String!) { disconnect(id: $id) }", {"id": id}
        )
